<a href="https://colab.research.google.com/github/vamshidhar23/2303A510H7-PYHTON-FOR-DATA-SCIENCE/blob/main/LAB_2_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Analyze student grades using NumPy arrays, including creating a 2D array, calculating statistics, using slicing and boolean indexing, reshaping the array, and commenting on the results.

## Create a 2d numpy array

### Subtask:
Create a 2D NumPy array to store student marks in 3 subjects.


**Reasoning**:
Create a 2D NumPy array with sample student grades and print it.



In [1]:
import numpy as np

student_grades = np.array([[85, 90, 78],
                           [92, 88, 95],
                           [78, 85, 80],
                           [90, 82, 87],
                           [88, 95, 91]])

print(student_grades)

[[85 90 78]
 [92 88 95]
 [78 85 80]
 [90 82 87]
 [88 95 91]]


## Calculate basic statistics

### Subtask:
Calculate and display the average, minimum, and maximum marks for all students and subjects.


**Reasoning**:
Calculate and print the average, minimum, and maximum marks for all students and subjects using NumPy functions.



In [2]:
average_mark = np.mean(student_grades)
min_mark = np.min(student_grades)
max_mark = np.max(student_grades)

print(f"Average mark for all students and subjects: {average_mark:.2f}")
print(f"Minimum mark for all students and subjects: {min_mark}")
print(f"Maximum mark for all students and subjects: {max_mark}")

Average mark for all students and subjects: 86.93
Minimum mark for all students and subjects: 78
Maximum mark for all students and subjects: 95


## Analyze student performance

### Subtask:
Use slicing to retrieve and display marks of specific students.


**Reasoning**:
Use slicing to select and display the marks of specific students as instructed.



In [3]:
# Select and display the marks of the first student
first_student_marks = student_grades[0, :]
print("Marks of the first student:")
print(first_student_marks)

# Select and display the marks of the third and fourth students
specific_students_marks = student_grades[2:4, :]
print("\nMarks of the third and fourth students:")
print(specific_students_marks)

Marks of the first student:
[85 90 78]

Marks of the third and fourth students:
[[78 85 80]
 [90 82 87]]


## Identify high-scoring students

### Subtask:
Use boolean indexing to find and display students who scored above 80 in any subject.


**Reasoning**:
Create a boolean array and use it to select rows where at least one value is greater than 80.



In [4]:
above_80_boolean = student_grades > 80
students_above_80 = student_grades[np.any(above_80_boolean, axis=1)]
print("Students who scored above 80 in any subject:")
print(students_above_80)

Students who scored above 80 in any subject:
[[85 90 78]
 [92 88 95]
 [78 85 80]
 [90 82 87]
 [88 95 91]]


## Analyze subject-wise performance

### Subtask:
Reshape the array to analyze performance across subjects.


**Reasoning**:
Reshape the array and calculate the average, minimum, and maximum marks for each subject.



In [5]:
subject_grades = student_grades.reshape(3, 5).T

average_subject_mark = np.mean(subject_grades, axis=0)
min_subject_mark = np.min(subject_grades, axis=0)
max_subject_mark = np.max(subject_grades, axis=0)

print("Average mark for each subject:", average_subject_mark)
print("Minimum mark for each subject:", min_subject_mark)
print("Maximum mark for each subject:", max_subject_mark)

Average mark for each subject: [86.6 85.6 88.6]
Minimum mark for each subject: [78 78 82]
Maximum mark for each subject: [92 95 95]


## Summarize findings

### Subtask:
Add comments to the code explaining the results of the analysis.


**Reasoning**:
The subtask is to add comments to the previously executed code cells to explain the results. I need to go back through the history of executed cells and add comments where calculations or filtering were performed, explaining the code and interpreting the output.



In [6]:
# Create a 2D NumPy array to store student marks in 3 subjects.
# Each row represents a student, and each column represents a subject.
student_grades = np.array([[85, 90, 78],
                           [92, 88, 95],
                           [78, 85, 80],
                           [90, 82, 87],
                           [88, 95, 91]])

# Print the created array to verify its structure and content.
print(student_grades)

# Calculate the average mark for all students and subjects using np.mean().
average_mark = np.mean(student_grades)
# Calculate the minimum mark across all students and subjects using np.min().
min_mark = np.min(student_grades)
# Calculate the maximum mark across all students and subjects using np.max().
max_mark = np.max(student_grades)

# Print the calculated statistics, formatted to two decimal places for the average.
# The output shows the overall performance: average=86.93, min=78, max=95.
print(f"Average mark for all students and subjects: {average_mark:.2f}")
print(f"Minimum mark for all students and subjects: {min_mark}")
print(f"Maximum mark for all students and subjects: {max_mark}")

# Use slicing to select and display the marks of the first student (index 0).
# The slice [0, :] selects the first row and all columns.
first_student_marks = student_grades[0, :]
print("Marks of the first student:")
print(first_student_marks) # Output: [85 90 78] - These are the marks of the first student in the three subjects.

# Use slicing to select and display the marks of the third and fourth students (indices 2 and 3).
# The slice [2:4, :] selects rows from index 2 up to (but not including) index 4, and all columns.
specific_students_marks = student_grades[2:4, :]
print("\nMarks of the third and fourth students:")
print(specific_students_marks) # Output: [[78 85 80] [90 82 87]] - These are the marks for the third and fourth students.

# Use boolean indexing to create a boolean array indicating where scores are above 80.
above_80_boolean = student_grades > 80
# Use np.any() with axis=1 to check if any score in a row (for a student) is True (above 80).
# This returns a boolean array where True indicates the student has at least one score above 80.
students_above_80 = student_grades[np.any(above_80_boolean, axis=1)]
# Print the rows (students) where at least one subject score was above 80.
# The output shows the students who meet this criterion.
print("Students who scored above 80 in any subject:")
print(students_above_80)

# Reshape the array to analyze performance across subjects.
# .reshape(3, 5) changes the array to 3 rows and 5 columns.
# .T (transpose) then makes it 5 rows and 3 columns, where each row is a subject and each column is a student's score in that subject.
# Note: This reshaping seems incorrect for analyzing performance *across* subjects where each row is a subject.
# A correct reshaping for subject analysis would likely have subjects as rows and students as columns.
# Let's assume the intent was to have subjects as rows and students as columns after reshaping and transposing.
# The original student_grades has students as rows and subjects as columns.
# To get subjects as rows and students as columns, we just need to transpose the original array.
# Let's correct the reshaping and analysis based on subjects as rows.
subject_grades_correct = student_grades.T

# Calculate the average mark for each subject using np.mean() with axis=1 (across columns, which are now students).
average_subject_mark = np.mean(subject_grades_correct, axis=1)
# Calculate the minimum mark for each subject using np.min() with axis=1.
min_subject_mark = np.min(subject_grades_correct, axis=1)
# Calculate the maximum mark for each subject using np.max() with axis=1.
max_subject_mark = np.max(subject_grades_correct, axis=1)

# Print the calculated statistics for each subject.
# The output shows the average, minimum, and maximum scores for each of the three subjects.
print("Average mark for each subject:", average_subject_mark)
print("Minimum mark for each subject:", min_subject_mark)
print("Maximum mark for each subject:", max_subject_mark)

[[85 90 78]
 [92 88 95]
 [78 85 80]
 [90 82 87]
 [88 95 91]]
Average mark for all students and subjects: 86.93
Minimum mark for all students and subjects: 78
Maximum mark for all students and subjects: 95
Marks of the first student:
[85 90 78]

Marks of the third and fourth students:
[[78 85 80]
 [90 82 87]]
Students who scored above 80 in any subject:
[[85 90 78]
 [92 88 95]
 [78 85 80]
 [90 82 87]
 [88 95 91]]
Average mark for each subject: [86.6 88.  86.2]
Minimum mark for each subject: [78 82 78]
Maximum mark for each subject: [92 95 95]


## Summary:

### Data Analysis Key Findings

*   The average mark across all students and subjects is 86.93, with a minimum mark of 78 and a maximum mark of 95.
*   The marks for the first student are 85, 90, and 78.
*   The marks for the third and fourth students are [[78, 85, 80], [90, 82, 87]].
*   All students scored above 80 in at least one subject.
*   The average marks for the three subjects are approximately 86.6, 88.0, and 86.2, respectively.
*   Subject 2 has the highest average mark (88.0).
*   The minimum marks for the three subjects are 78, 82, and 78, respectively.
*   The maximum marks for the three subjects are 92, 95, and 95, respectively.

### Insights or Next Steps

*   Analyze the performance of individual students to identify those who may need additional support or those who are excelling.
*   Investigate the subject with the lowest average score (Subject 3) to understand potential reasons for lower performance.
